yew/src/virtual_dom/vnode.rs

```diff
@@ -32,13 +33,31 @@ pub enum VNode {
 }
 
 impl VNode {
-    pub fn key(&self) -> &Option<String> {
+    pub fn key(&self) -> &Option<Key> {
         match self {
-            VNode::VTag(vtag) => &vtag.key,
-            VNode::VText(_) => &None,
             VNode::VComp(vcomp) => &vcomp.key,
             VNode::VList(vlist) => &vlist.key,
             VNode::VRef(_) => &None,
+            VNode::VTag(vtag) => &vtag.key,
+            VNode::VText(_) => &None,
+        }
+    }
+
+    pub fn reference(&self) -> Option<Node> {
```

Comment:
What if this returned `Node` instead of an option?

Reply:
> I'm worried that the current implementation of reference() doesn't return a reference for VList and I'm not super confident that VComp will always have a reference when it should either.

I think I am missing something in the logic of `VComp`s, because if they are always in `Unmounted` state when calling `apply` (which I think is the case), then they will always have a node: the dummy text one or the ancestor's one.

We need to clarify this before changing the return type to `Node`. But indeed I would like this to not return an option.

Ah and also, what should we return for `VList`s? IMO returning the first child won't work because moving the first child won't move the entire `VList`.

Reply:
```rust
pub fn reference(&self) -> Either<Node, Vec<Node>>
```
?

Reply:
> Ah and also, what should we return for VLists? IMO returning the first child won't work because moving the first child won't move the entire VList.

I think first child will work because this is only used for `next_sibling`, it's not used for `ancestor`.

Reply:
But maybe we should call it something other than `reference`. Maybe `marker_node` or something like that?

Reply:
I was thinking about this:
https://github.com/yewstack/yew/blob/8fd55baf7642e7453c02e0fd032e5cbf6455597f/yew/src/virtual_dom/vlist.rs#L414-L433
I don't think it would works with `marker_node`.

Reply:
Oh when you need to move a list?

Reply:
Yes, if the list is keyed and moved, we need to move all its children (since it doesn't exist itself in the DOM).

Reply:
got it, `Either<Node, Vec<Node>>` sounds good to me!